Fix HPA feedback from writing status.replicas to spec.replicas.

josephburnett opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

> /kind bug

**What this PR does / why we need it**:

There are various reasons that the HPA will decide not the change the current scale.  Two important ones are when missing metrics might change the direction of scaling, and when the recommended scale is within tolerance of the current scale.

The way that `ReplicaCalculator` signals it's desire to not change the current scale is by returning the current scale.  However the current scale is from `scale.Status.Replicas` and can be larger than `scale.Spec.Replicas` (e.g. during Deployment rollout with configured surge).  This causes a positive feedback loop because `scale.Status.Replicas` is written back into `scale.Spec.Replicas`, further increasing the current scale.

This PR fixes the feedback loop by plumbing the replica count from spec through `horizontal.go` and `replica_calculator.go` so the calculator can punt with the right value.

It also introduces separate types for replica counts derived from `scale.Spec` and `scale.Status` (`specReplicas` and `statusReplicas` respectively) to guard against this kind of cross-talk.  When returning a desired scale to be written into spec, the calculator must either return the given, current scale, or explicitly call `newSpecReplicas`.

With separate types, other sources of cross-talk became compiler errors.  E.g. [recording Status.Replicas as an initial recommendation](https://github.com/kubernetes/kubernetes/blob/cb862e332329736dc1b893da6f757b808fa802aa/pkg/controller/podautoscaler/horizontal.go#L573-L574).  This would manifest if a deployment was in the process of rolling out when the HPA reboots.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #78712
Fixes #72775

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

mwielgus commented:
/ok-to-test

josephburnett commented:
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.

New comment.

josephburnett commented:
> If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.

Done.

josephburnett commented:
/test pull-kubernetes-integration

josephburnett commented:
/test pull-kubernetes-verify

josephburnett commented:
/test pull-kubernetes-verify

josephburnett commented:
/test pull-kubernetes-integration

josephburnett commented:
[/batman](https://www.youtube.com/watch?v=44-RsrF_V_w)

soggiest commented:
@kubernetes/sig-autoscaling-pr-reviews Can someone take a look at this
@claurence Do you think this can still make it into the release of should we punt?

josephburnett commented:
@krzysztof-jastrzebski could you also take a look, since you dug into this a little bit.  Did I address all the defects which you noticed?  Thanks!

Once this lands, I will want to patch it into all the recent major versions.

mwielgus commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -229,13 +229,28 @@ func (a *HorizontalController) processNextWorkItem() bool {
+// Types to help keep status and spec separate and avoid a positive
+// feedback loop.  Generally, nothing from status should go back into
+// spec.
+type specReplicas int32
```
Where does specReplicas go? To Deployment/RS specification (scale subresource)?

josephburnett replied:
specReplicas goes to the scale subresource (spec).

Re: statusReplicas, I removed it from all the places it wasn't needed.  I tried a few different approaches to fixing this problem and that plumbing was just a leftover.

mwielgus reviewed (changes requested):
In multiple places in the code you pass statusReplicas as a function argument but you make no use of it. They are not needed by any interface or so. At least for me, it is confusing why would you need to ever pass statusReplicas to a function that gets a metrics. 
Please remove these parameters.

mwielgus commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -229,13 +229,27 @@ func (a *HorizontalController) processNextWorkItem() bool {
+	metricSpecs []autoscalingv2.MetricSpec) (replicas specReplicas, metric string, statuses []autoscalingv2.MetricStatus, timestamp time.Time, err error) {
 
-	currentReplicas := scale.Status.Replicas
+	specReplicas := specReplicas(scale.Spec.Replicas)
```
What implication does this change has?

mwielgus replied:
Nevermind, it's explained in the PR description :).

josephburnett replied:
This is the root of the bug.  Writing status back into spec.

mwielgus commented on pkg/controller/podautoscaler/replica_calculator_test.go:
```diff
@@ -972,8 +1066,30 @@ func TestReplicaCalcToleranceCMObject(t *testing.T) {
+func TestReplicaCalcTolerancePerPodCMObjectSurge(t *testing.T) {
+	tc := replicaCalcTestCase{
+		currentSpecReplicas:   4,
+		currentStatusReplicas: 8,
```
What would happen if we had 0 in either spec or status (or both)?

josephburnett replied:
These test case fields are misleading.  We don't do anything with `statusReplicas` except write it down any more.  So a zero value is uninteresting.  Likewise, `specReplicas` is either left unmodified (in which case no big deal) or set to the desired scale.  In this case, zero already has the meaning, don't scale.  So nothing will happen anyway.  We already have a unit tests for this: https://github.com/kubernetes/kubernetes/blob/9120d427dab4a69ddcf22597f81b574fb1af96bc/pkg/controller/podautoscaler/horizontal_test.go#L1633

I've renamed these as `specReplicas` and `actualPodCount`.  The actual pod count is just used by the test to setup mocks.

mwielgus commented:
Please describe what manual tests did you perform on this change. In particular, please confirm that you built a Kubernetes version with this PR included, deployed it, created HPA and made sure it doesn't crash.

josephburnett commented:
> Please describe what manual tests did you perform on this change. In particular, please confirm that you built a Kubernetes version with this PR included, deployed it, created HPA and made sure it doesn't crash.

The PR test automation includes a least one HPA e2e test, so I did no manual testing:

![image](https://user-images.githubusercontent.com/1103629/59604584-dde63f00-910c-11e9-8cc6-062f14ea1cf1.png)

mwielgus commented:
I know that there are e2e test. AFAIR I wrote some of them ;). The thing is that we are well in the code freeze and the PR is quite large. I want to be double sure that we won't have to rollback or send subsequent last minute patches or postpone the K8S release. Not everything is checked by e2e and an eye of an engineer can make a difference. Please do the build :).

krzysztof-jastrzebski commented:
I think it would be easier to review if you split PR into:
1. PR adding specReplica type and all refactors.
2. PR fixing the bug.

krzysztof-jastrzebski commented:
I think it would be easier to review if you split PR into:
1. PR adding specReplica type and all refactors.
2. PR fixing the bug.

josephburnett commented:
> easier to review if you split PR

@krzysztof-jastrzebski yes, I think you're right.  I started with the type change and let the compiler guide me to all the places we did it wrong.  But now that I know exactly where, I could pull out a more surgical fix.

@mwielgus, ack.  I'll start working on a manual build and test.  ETA tomorrow.

gliush commented on pkg/controller/podautoscaler/replica_calculator.go:
```diff
@@ -252,26 +252,26 @@ func (c *ReplicaCalculator) GetObjectMetricReplicas(currentReplicas int32, targe
+		replicaCount = newSpecReplicas(int32(math.Ceil(float64(utilization) / float64(targetAverageUtilization))))
 	}
-	utilization = int64(math.Ceil(float64(utilization) / float64(currentReplicas)))
+	utilization = int64(math.Ceil(float64(utilization) / float64(specReplicas)))
```
I'm not sure that this is a good idea.
I mean, scale.Status.Replicas and scale.Spec.Replicas might change a lot.
If I'm not mistaken these numbers might differ a lot. And in this case the utilization here ^^ will be calculated incorrectly.

josephburnett replied:
You are correct.  I'm using the wrong denominator.  I'll plumb statusReplicas back through.

josephburnett replied:
And my use of `replicaCount` (which is `specReplicas`) when calculating the original `usageRatio` is wrong too.  I don't think `specReplicas` should ever be in the denominator of a ratio calculation because it doesn't relate directly to the metrics and pods found.  Does that sound right?

josephburnett replied:
@gliush I've been thinking about the per-pod object metrics use case.  And trying to make the unit tests make sense.

If I divide by `statusReplicas` then the utilization changes enough to exceed the tolerance and enter the conditional to update `replicaCount`.  However the calculation divides the object metric by the configured target, not taking replica counts into account.  So it will always come up with the `statusReplicas` (and maybe +1 because of the ceiling).  There is nothing else for it to do because adding or removing Pods will not change the object metric.

So dividing by `statusReplicas` is pointless.  In this particular case, I think dividing by `statusReplicas` is the right thing to do.

gliush replied:
>  I don't think specReplicas should ever be in the denominator of a ratio calculation because it doesn't relate directly to the metrics and pods found. 

Right.

> However the calculation divides the object metric by the configured target, not taking replica counts into account. 

It does take it into account. Let me show it:
1. `usageRatio := float64(utilization) / float64(targetUtilization)`
https://github.com/kubernetes/kubernetes/blob/b8b1720f127f9a4e6784bc652faafb256322353f/pkg/controller/podautoscaler/replica_calculator.go#L243

2. `replicaCount = int32(math.Ceil(usageRatio * float64(readyPodCount)))`
https://github.com/kubernetes/kubernetes/blob/b8b1720f127f9a4e6784bc652faafb256322353f/pkg/controller/podautoscaler/replica_calculator.go#L255

3. `return replicaCount ...`

I.e. the `utilizationRatio` is indeed calculated without the current number of pods. But the desired number of pods is calculated by multiplying the current number of pods by that `utilizationRatio`. That is where the current `statusReplicas` is used.

Another argument, more conceptual.
The idea of HPA itself is to autoscale pods under control of some metric value.
It means that we should always consider current load to be influenced (formed) by a current number of pods. And that increased number of pods will reduce the metric value.

As a result, it doesn't make sense to use the `specReplicas` in any calculation, because current state depends only on `statusReplicas`.

josephburnett replied:
Your example above is a link to `GetObjectMetricReplicas` which does indeed consider pod count.  So of course we should divide by `statusReplicas` or `readyPodCount`.

https://github.com/kubernetes/kubernetes/blob/b8b1720f127f9a4e6784bc652faafb256322353f/pkg/controller/podautoscaler/replica_calculator.go#L255

However your initial comment and the interesting case here is in `GetObjectPerPodMetricReplicas`.  We don't divide by the number of pods.  Just the metric by the target.

https://github.com/kubernetes/kubernetes/blob/b8b1720f127f9a4e6784bc652faafb256322353f/pkg/controller/podautoscaler/replica_calculator.go#L272

claurence commented:
@josephburnett this item is tagged against 1.15 release - our release date is today - is this issue release blocking in the opinion of the SIGs impacted?  It doesn't sound like it is fixing a regression - just fixing an issue in 1.14 - can you let us know why this issue can't go in either a 1.15.x patch or 1.16?

tpepper commented:
@janetkuo @kow3ns @mattfarina @prydonius looking for confirmation here ASAP whether this is 1.15 release blocking since it was added to the milestone and release is today.

kow3ns commented:
@tpepper imo sig-autoscaling not sig-apps should be the decider for blocking the release. However, my personal opinion FWIW is that this is not a release blocking.

tpepper commented:
@mwielgus can you confirm whether this should be treated as 1.15 release blocking or not?

tpepper commented:
> @mwielgus can you confirm whether this should be treated as 1.15 release blocking or not?

From slack:
#79035 is not a release blocker. It fixes a problem that has been there for a while. It can go in the first patch release.

Once this PR merges to master, please open a [cherry-pick](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) against release-1.15 branch.

Bubblemelon commented:
As discussed on https://kubernetes.slack.com/archives/C18NZM5K9/p1560793182057800 , this is not release blocking.

gliush commented:
@josephburnett : I don't like the idea when replicaCalculator returns "the same replica number" at all.
I believe we should differentiate the "number of replicas" int32 and "if we should change the number of replicas" boolean.

josephburnett commented:
> I believe we should differentiate the "number of replicas" int32 and "if we should change the number of replicas" boolean.

That was my first approach.  But it substantially increased the amount of refactoring necessary because most callers throw the value right into variable to return.  Adds a lot of checks.  E.g. https://github.com/kubernetes/kubernetes/blob/eaf89cfbb483f99411141f70c294e8a014e49cec/pkg/controller/podautoscaler/horizontal.go#L285-L330

josephburnett commented:
@krzysztof-jastrzebski and @mwielgus, I have reverted all the type changes to get a minimal diff for patching.  I'll submit the type safety changes as a separate pull request after.  This minimal change is the one I will test manually.

josephburnett commented:
/test pull-kubernetes-integration

krzysztof-jastrzebski commented:
You need to change also reconcileAutoscaler function as it also uses status.Replicas. 
I think we should never scale up based on:
desiredReplicas = a.normalizeDesiredReplicas(hpa, key, currentReplicas, desiredReplicas)
If someone changes number of replicas manually for a moment then HPA can store the value and scale up to it for 5 minute even if user manually scales down. This mechanism was to slow down scale down not do scale ups.
Also rescaleReason probably should be updated if 
desiredReplicas = a.normalizeDesiredReplicas(hpa, key, currentReplicas, desiredReplicas)
changes desiredReplicas.

gliush commented on pkg/controller/podautoscaler/replica_calculator.go:
```diff
@@ -259,19 +259,19 @@ func (c *ReplicaCalculator) GetObjectMetricReplicas(currentReplicas int32, targe
 	}
 
-	replicaCount = currentReplicas
+	replicaCount = specReplicas
```
It is not clear to me why we didn't use
```
readyPodCount, err := c.getReadyPodsCount(namespace, podSelector
```
as we do in all other functions in this file.

gliush replied:
I mean, the "currentReplicas" and "readyReplicas" may differ.

josephburnett replied:
Yes, it looks weird.  But I think the answer is pod count doesn't matter, whether they are ready or not.

See my explanation above: https://github.com/kubernetes/kubernetes/pull/79035#discussion_r294995711

josephburnett commented:
> You need to change also reconcileAutoscaler function as it also uses status.Replicas.

This is part of the larger pull request which uses types to make sure I don't do something stupid like this.  However when I stripped out the types, I accidentally reverted this fix.

https://github.com/josephburnett/kubernetes/blob/1f1461887b7de67616ab18de953bdd418472196f/pkg/controller/podautoscaler/horizontal.go#L60

and

https://github.com/josephburnett/kubernetes/blob/1f1461887b7de67616ab18de953bdd418472196f/pkg/controller/podautoscaler/horizontal.go#L588-L589

Good catch @krzysztof-jastrzebski!  I've added a commit to apply that change again.

josephburnett commented:
@mwielgus I have good new and bad news.  This PR appears to indeed fix the issues with positive feedback from writing status replicas to spec replicas.  I've patched this minimal PR without additional types into GKE 1.12.9 and ran the patched controller manager in a GKE 1.12.9 cluster.  I put a deployment under load and then removed it.  It scaled up and then down as expected, so I think this PR is good-to-go.  And I can continue patching other GKE major versions with the same fix.

However, I have discovered another bug.  When a Deployment is comprised of a sufficient number of Pending unscheduled pods (e.g. due to lack of cluster capacity) we generate enough false 100% metrics to keep the utilization ratio at 1.0 or higher.  Actual utilization is 0 because load has been removed.  With an apparent change in direction due to the missing metrics, the autoscaler punts and returns the current scale.  So we never scale down, even when utilization is 0%.

I ran into this when testing a Deployment rollout that created a lot of Pending pods.  However I think I will address that issue in a separate PR.

josephburnett commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

josephburnett commented:
Also, we should definitely write an E2E test for this.  Deploying new code under load without the HPA losing it's mind is an important use case.

josephburnett commented:
@mwielgus I have run one additional manual test.  I scaled the cluster up to 10 nodes to have sufficient capacity and then made several changes to the Deployment under load.  In each case, the Deployment made additional pods (above the HPA max of 10) but the HPA didn't enter the feedback loop.  So again, I think this PR is good-to-go.

Re: the stuck HPA on missing metrics bug (https://github.com/kubernetes/kubernetes/pull/79035#issuecomment-503326840), I have opened a separate issue: https://github.com/kubernetes/kubernetes/issues/79158 and I'll work on that next.

gliush commented:
I still believe that we should use only "ReadyPods" in all the calculations.
Consider the following example that I run on my `minikube`:

```$ cat /tmp/d.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  replicas: 100
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx
        name: nginx
        resources:
          limits:
            cpu: "1000m"
          requests:
            cpu: "1000m"

$ kubectl create -f /tmp/d.yaml

$ kubectl get deployments nginx -o=jsonpath='{.spec.replicas}{"\t"}{.status.replicas}{"\n"}'
100 100

$ kubectl get pods | grep "Running" | wc -l 
2
```

So, I have 100 pods in Spec and in Status, while I have only 2 of them running. (In minikube I have only 2 cpus, but it may be caused by plenty of different reasons).

All other pods are in “Pending” state
If metrics show that we need to double the number of pods, HPA controller will set “Replicas: 200”. 

While only 2 pods are working, and in reality we need 4 of them

The next `reconcile` will increase it to “400” again, because the real number of pods was not changed, and the metric won’t change. 

This is the main problem with “specReplicas”, I believe. That it is not connected to reality.

josephburnett commented:
> I still believe that we should use only "ReadyPods" in all the calculations.

@gliush and I had a conversation about this behavior and came to agreement.  In particular the difference between specifying `Value` or `AverageValue` and how the implementation treats the current number of Pods.  I believe that the correct behavior is to divide by `specReplicas` not by `statusReplicas` or ready pods.  That's because specifying an `AverageValue` doesn't put the autoscaler into a feedback loop like `Value`--it just reconciles to the current metric value by the target value.  This is not at all clear from the documentation, so I've opened https://github.com/kubernetes/kubernetes/issues/79272 to clarify.

However since the current implementation is using `statusReplicas` when calculating the `usageRatio` (by accident I think) I will change this PR to do the same.  That means we needlessly update `replicaCount` whenever `statusReplicas` exceeds `specReplicas` (e.g. a Deployment rollout is going on).  But the effect is minimal.  I will open a separate PR to fix that bug separately, dividing by `specReplicas` instead.

josephburnett commented:
/test pull-kubernetes-integration
/test pull-kubernetes-bazel-test

krzysztof-jastrzebski commented on pkg/controller/podautoscaler/replica_calculator.go:
```diff
@@ -334,7 +334,7 @@ func (c *ReplicaCalculator) GetExternalMetricReplicas(currentReplicas int32, tar
 // target metric value per pod (as a milli-value) for the external metric in the
 // given namespace, and the current replica count.
-func (c *ReplicaCalculator) GetExternalPerPodMetricReplicas(currentReplicas int32, targetUtilizationPerPod int64, metricName, namespace string, metricSelector *metav1.LabelSelector) (replicaCount int32, utilization int64, timestamp time.Time, err error) {
+func (c *ReplicaCalculator) GetExternalPerPodMetricReplicas(statusReplicas int32, targetUtilizationPerPod int64, metricName, namespace string, metricSelector *metav1.LabelSelector) (replicaCount int32, utilization int64, timestamp time.Time, err error) {
```
Why "statusReplica". Is increasing number of pods during rolling upgrade increasing external metric? Additional pods  added during upgrade shouldn't serve as if they are healthy then old pods are deleted.

josephburnett replied:
Incidentally, it doesn't make a difference whether we use status or spec replicas here.  It's technically a bug to use status replicas, but I left it as-is in order to fix one bug at a time.  https://github.com/kubernetes/kubernetes/issues/79274 tracks fixing the bug to use spec replicas.

josephburnett replied:
Also, it's super confusing that we pass in spec replicas at all.  We only use that value in order to signal a decision to not scale by returning the current desired replica count.  Better would be to return a boolean.  https://github.com/kubernetes/kubernetes/issues/79275 tracks that improvement.

krzysztof-jastrzebski commented:
/lgtm

MaciekPytel commented:
/approve
Great stuff, sorry to be so unresponsive on review.

josephburnett commented:
/kind bug

MaciekPytel commented:
/kind bug

josephburnett commented:
I need someone who can manually remove the needs-kind label.

![image](https://user-images.githubusercontent.com/1103629/60509334-f3429800-9ccc-11e9-8cbc-c06bd3620ebc.png)

josephburnett commented:
/hold

I would like to squash these commits.

Commit:
There are various reasons that the HPA will decide not the change the
current scale. Two important ones are when missing metrics might
change the direction of scaling, and when the recommended scale is
within tolerance of the current scale.

The way that ReplicaCalculator signals it's desire to not change the
current scale is by returning the current scale. However the current
scale is from scale.Status.Replicas and can be larger than
scale.Spec.Replicas (e.g. during Deployment rollout with configured
surge). This causes a positive feedback loop because
scale.Status.Replicas is written back into scale.Spec.Replicas,
further increasing the current scale.

This PR fixes the feedback loop by plumbing the replica count from
spec through horizontal.go and replica_calculator.go so the calculator
can punt with the right value.

josephburnett commented:
/hold cancel.

I've squashed and force pushed all the commits in this PR.  The original commits can be found in my repo: https://github.com/josephburnett/kubernetes/compare/hparunaway-base...josephburnett:hparunaway-original

josephburnett commented:
/hold cancel

krzysztof-jastrzebski commented:
/lgtm

juanpmarin commented:
Hi, is this currently released ?

austinpray commented:
> Hi, is this currently released ?

@juanpmarin it looks like this is gonna be in 1.16+

```
λ  kubernetes master ✓ git tag --contains 39c4875321991f305d51e30481a66701b6b76f5f
v1.16.0-alpha.1
v1.16.0-alpha.2
v1.16.0-alpha.3
v1.16.0-beta.0
v1.16.0-beta.1
v1.17.0-alpha.0
```

kimxogus commented:
@juanpmarin This is also cherry picked to 1.13(#79707), 1.14(#79708), 1.15(#79709). Latest version of them may have this fix too.

paalkr commented:
I would just like to draw some attention to https://github.com/kubernetes/kubernetes/issues/72775. Many are still facing issues with rolling updates.
